### DISCUSSION TITLE
原生 OpenWrt 编译运行支持

### DISCUSSION THREAD

**Discussion by 1715173329** [XIU2/CloudflareSpeedTest#64] [分享]
~~项目地址：[immortalwrt-collections/openwrt-cdnspeedtest](https://github.com/immortalwrt-collections/openwrt-cdnspeedtest)~~
由于维护问题，不再单独建立一个仓库，如有需要可以参考：[immortalwrt/packages/cdnspeedtest](https://github.com/immortalwrt/packages/tree/master/net/cdnspeedtest)

_* 适用于官方 OpenWrt **19.07** 及以上版本，以及国内大多数修改分支_

使用方法
- ~~从 [Releases](https://github.com/immortalwrt-collections/openwrt-cdnspeedtest/releases) 下载对应的预编译软件包~~
- 从 [ImmortalWrt 下载站](https://downloads.immortalwrt.org/) 下载对应的预编译软件包
- 自行编译
  ```bash
  # 进入 OpenWrt buildroot 根目录
  $ mkdir -p "package/cdnspeedtest"
  $ curl -fL "https://raw.githubusercontent.com/immortalwrt/packages/master/net/cdnspeedtest/Makefile" | sed 's,../../lang,$(TOPDIR)/feeds/packages/lang,' > "package/cdnspeedtest/Makefile"

  # 进入 buildroot menuconfig 配置界面
  $ make menuconfig
  # 选中 Network -> cdnspeedtest
  Network  --->
  <*> cdnspeedtest.............. Getting the fastest ips to your network of CDN
  ```

文件结构：
```
- /usr/bin/
	- cdnspeedtest # 核心程序
- /usr/share/CloudflareSpeedTest/
	- ip.txt # IPv4 地址数据
	- ipv6.txt # IPv6 地址数据
```

**Comment by XIU2** [comment]
看起来不错，不过我没有 OpenWrt 设备，对这些也不懂。。。

**Comment by loofeal** [comment]


**Reply by 1715173329** [reply]
放入 openwrt 里的 packages 文件夹，然后 menuconfig 选中后编译即可

**Reply by openwrt-try** [reply]
这个似乎没有luci界面？没有的话是不是只能命令的形式运行了，小白不懂，如果没有能否请大神加入luci界面，

**Reply by 1715173329** [reply]
本身就是命令行工具，，

**Reply by luoxufeiyan** [reply]
@openwrt-try 发现了一个LuCI 的带界面的项目： https://github.com/mingxiaoyu/luci-app-cloudflarespeedtest

**Comment by loofeal** [comment]
可以试试

**Comment by 1715173329** [comment]
做了个 CI，会自动构建常见架构的软件包，如果有其他架构需要支持欢迎开 issue。
https://github.com/immortalwrt-collections/openwrt-cdnspeedtest/releases

**Comment by sumulige** [comment]
这个编译和opkg install cdnspeedtest_2.0.3-1_x86_64.ipk 是一个意思么

**Comment by rheatin** [comment]
很需要这个，谢谢天灵大大，和作者哈哈哈

**Reply by wutaijieing** [reply]
怎么对接推送？

**Comment by kidxiang** [comment]
京东云亚瑟AX1800pro正好需要，感谢

**Comment by snakwu** [comment]
> 项目地址：[immortalwrt-collections/openwrt-cdnspeedtest](https://github.com/immortalwrt-collections/openwrt-cdnspeedtest) _* 基于 **[GNU General Public License v3.0](https://github.com/immortalwrt-collections/openwrt-cdnspeedtest/blob/master/LICENSE)** 协议进行分发_ _* 适用于官方 OpenWrt **19.07** 及以上版本，以及国内大多数修改分支_
> 
> 使用方法
> 
> * 从 [Releases](https://github.com/immortalwrt-collections/openwrt-cdnspeedtest/releases) 下载对应的预编译软件包
> * 自行编译
>   ```shell
>   # 进入 OpenWrt buildroot 根目录
>   $ echo "src-git cdnspeedtest https://github.com/immortalwrt-collections/openwrt-cdnspeedtest.git" >> "feeds.conf.default"
>   $ ./scripts/feeds update -a
>   $ ./scripts/feeds install golang cdnspeedtest
>   
>   # 进入 buildroot menuconfig 配置界面
>   $ make menuconfig
>   # 选中 Network -> cdnspeedtest
>   Network  --->
>   <*> cdnspeedtest.............. Getting the fastest ips to your network of CDN
>   ```
> 
> 文件结构：
> 
> ```
> - /usr/bin/
> 	- cdnspeedtest # 核心程序
> - /usr/share/CloudflareSpeedTest/
> 	- ip.txt # IPv4 地址数据
> 	- ipv6.txt # IPv6 地址数据
> ```

官方源码，添加无法编译，请问怎么解决？谢谢
make[3]: Entering directory '/home/snakwu/openwrt/openwrt/package/snakwu/openwrt-cdnspeedtest/cdnspeedtest'
rm -f /home/snakwu/openwrt/openwrt/build_dir/target-x86_64_musl/CloudflareSpeedTest-2.2.5/.built
touch /home/snakwu/openwrt/openwrt/build_dir/target-x86_64_musl/CloudflareSpeedTest-2.2.5/.built_check
CONFIG_GOLANG_MOD_CACHE_WORLD_READABLE="" GO_BUILD_CACHE_DIR="/home/snakwu/openwrt/openwrt/tmp/go-build" GO_MOD_CACHE_DIR="/home/snakwu/openwrt/openwrt/dl/go-mod-cache" GO_MOD_ARGS="-modcacherw" GO_PKG="github.com/XIU2/CloudflareSpeedTest" GO_INSTALL_EXTRA="" GO_INSTALL_ALL="" GO_SOURCE_ONLY="" GO_BUILD_PKG="github.com/XIU2/CloudflareSpeedTest/..." GO_EXCLUDES="" GO_GO_GENERATE="" GO_INSTALL_BIN_PATH="/usr/bin" BUILD_DIR="/home/snakwu/openwrt/openwrt/build_dir/target-x86_64_musl/CloudflareSpeedTest-2.2.5" GO_BUILD_DIR="/home/snakwu/openwrt/openwrt/build_dir/target-x86_64_musl/CloudflareSpeedTest-2.2.5/.go_work/build" GO_BUILD_BIN_DIR="/home/snakwu/openwrt/openwrt/build_dir/target-x86_64_musl/CloudflareSpeedTest-2.2.5/.go_work/build/bin" GO_BUILD_DEPENDS_PATH="/usr/share/gocode" GO_BUILD_DEPENDS_SRC="/home/snakwu/openwrt/openwrt/staging_dir/target-x86_64_musl/usr/share/gocode/src" GOOS="linux" GOARCH="amd64" GO386="" GOAMD64="v1" GOARM="" GOMIPS="" GOMIPS64="" GOPPC64="" CC="x86_64-openwrt-linux-musl-gcc" CXX="x86_64-openwrt-linux-musl-g++" CGO_CFLAGS="-Os -pipe -fno-caller-saves -fno-plt -fhonour-copts -fmacro-prefix-map=/home/snakwu/openwrt/openwrt/build_dir/target-x86_64_musl/CloudflareSpeedTest-2.2.5=CloudflareSpeedTest-2.2.5 -Wformat -Werror=format-security -fstack-protector -D_FORTIFY_SOURCE=1 -Wl,-z,now -Wl,-z,relro" CGO_CPPFLAGS="-I/home/snakwu/openwrt/openwrt/staging_dir/toolchain-x86_64_gcc-13.3.0_musl/usr/include -I/home/snakwu/openwrt/openwrt/staging_dir/toolchain-x86_64_gcc-13.3.0_musl/include -I/home/snakwu/openwrt/openwrt/staging_dir/toolchain-x86_64_gcc-13.3.0_musl/include/fortify" CGO_CXXFLAGS="-Os -pipe -fno-caller-saves -fno-plt -fhonour-copts -fmacro-prefix-map=/home/snakwu/openwrt/openwrt/build_dir/target-x86_64_musl/CloudflareSpeedTest-2.2.5=CloudflareSpeedTest-2.2.5 -Wformat -Werror=format-security -fstack-protector -D_FORTIFY_SOURCE=1 -Wl,-z,now -Wl,-z,relro" CGO_LDFLAGS="-L/home/snakwu/openwrt/openwrt/staging_dir/toolchain-x86_64_gcc-13.3.0_musl/usr/lib -L/home/snakwu/openwrt/openwrt/staging_dir/toolchain-x86_64_gcc-13.3.0_musl/lib -fuse-ld=bfd -znow -zrelro" CGO_ENABLED=0 GOPATH="/home/snakwu/openwrt/openwrt/build_dir/target-x86_64_musl/CloudflareSpeedTest-2.2.5/.go_work/build" GOCACHE="/home/snakwu/openwrt/openwrt/tmp/go-build" GOMODCACHE="/home/snakwu/openwrt/openwrt/dl/go-mod-cache" GOENV=off GOTOOLCHAIN=local /usr/bin/env bash /home/snakwu/openwrt/openwrt/feeds/packages/lang/golang//golang-build.sh build -v -buildvcs=false -trimpath -ldflags "all=-buildid '1723576033' -linkmode external -extldflags '-L/home/snakwu/openwrt/openwrt/staging_dir/toolchain-x86_64_gcc-13.3.0_musl/usr/lib -L/home/snakwu/openwrt/openwrt/staging_dir/toolchain-x86_64_gcc-13.3.0_musl/lib -fuse-ld=bfd -Wl,-z,now -Wl,-z,relro'"        -ldflags " -X main.version=v2.2.5 -buildid '1723576033' -linkmode external -extldflags '-L/home/snakwu/openwrt/openwrt/staging_dir/toolchain-x86_64_gcc-13.3.0_musl/usr/lib -L/home/snakwu/openwrt/openwrt/staging_dir/toolchain-x86_64_gcc-13.3.0_musl/lib -fuse-ld=bfd -Wl,-z,now -Wl,-z,relro'"  
Finding targets

Building targets
-linkmode requires external (cgo) linking, but cgo is not enabled

make[3]: *** [Makefile:54: /home/snakwu/openwrt/openwrt/build_dir/target-x86_64_musl/CloudflareSpeedTest-2.2.5/.built] Error 1
make[3]: Leaving directory '/home/snakwu/openwrt/openwrt/package/snakwu/openwrt-cdnspeedtest/cdnspeedtest'
time: package/snakwu/openwrt-cdnspeedtest/cdnspeedtest/compile#0.45#0.67#4.10
    ERROR: package/snakwu/openwrt-cdnspeedtest/cdnspeedtest failed to build.
make[2]: *** [package/Makefile:177: package/snakwu/openwrt-cdnspeedtest/cdnspeedtest/compile] Error 1
make[2]: Leaving directory '/home/snakwu/openwrt/openwrt'
make[1]: *** [package/Makefile:171: /home/snakwu/openwrt/openwrt/staging_dir/target-x86_64_musl/stamp/.package_compile] Error 2
make[1]: Leaving directory '/home/snakwu/openwrt/openwrt'
make: *** [/home/snakwu/openwrt/openwrt/include/toplevel.mk:248: world] Error 2
snakwu@compiler-server:~/openwrt/openwrt$
